packages/teko/src/Tpetra/Teko_BlockingTpetra.cpp

```diff
@@ -164,148 +165,193 @@ void many2one(Tpetra::MultiVector<ST, LO, GO, NT>& one,
+  auto values =
+      Kokkos::View<ST*>(Kokkos::ViewAllocateWithoutInitializing("values"), totalNumOwnedCols);
+
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<Kokkos::Schedule<Kokkos::Dynamic>, matrix_execution_space>(0, numMyRows),
+      KOKKOS_LAMBDA(const LO localRow) {
+        GO globalRow             = gRowMap_dev.getGlobalElement(localRow);
+        LO lid                   = A_rowmap_dev.getLocalElement(globalRow);
+        const auto sparseRowView = A_dev.row(lid);
+
+        LO colId      = 0;
+        LO colIdStart = prefixSumEntriesPerRow[localRow];
+        for (auto localCol = 0; localCol < sparseRowView.length; localCol++) {
+          GO gid = data(sparseRowView.colidx(localCol), 0);
+          if (gid == invalid) continue;
+          auto value                              = sparseRowView.value(localCol);
+          uniqueColumnIndices(colId + colIdStart) = gid;
+          values(colId + colIdStart)              = value;
+          colId++;
+        }
+      });
+
+  Kokkos::deep_copy(columnIndices, uniqueColumnIndices);
+  auto exec_space = matrix_execution_space{};
+  Kokkos::sort(uniqueColumnIndices);
+  auto end = Kokkos::Experimental::unique(exec_space, uniqueColumnIndices);
+  auto nUnique =
+      Kokkos::Experimental::distance(Kokkos::Experimental::begin(uniqueColumnIndices), end);
+  auto listColumnIndices = Kokkos::subview(uniqueColumnIndices, Kokkos::pair<int, int>(0, nUnique));
+
+  auto colMap = RCP<Tpetra::Map<LO, GO, NT>>(new Tpetra::Map<LO, GO, NT>(
+      invalid, listColumnIndices, rangeMap->getIndexBase(), rangeMap->getComm()));
+
+  auto colMap_dev         = colMap->getLocalMap();
+  auto localColumnIndices = Kokkos::View<LO*>(
+      Kokkos::ViewAllocateWithoutInitializing("localColumnIndices"), totalNumOwnedCols);
+  Kokkos::parallel_for(
+      team_policy(numMyRows, Kokkos::AUTO()), KOKKOS_LAMBDA(const team_member& t) {
+        const auto localRow = t.league_rank();
+        const auto numCols  = nEntriesPerRow(localRow);
+        const auto colStart = prefixSumEntriesPerRow(localRow);
+        for (auto localCol = 0; localCol < numCols; localCol++) {
+          const auto globalColId                  = columnIndices(localCol + colStart);
+          const auto localColId                   = colMap_dev.getLocalElement(globalColId);
+          localColumnIndices(localCol + colStart) = localColId;
+        }
+        auto rowColumnIndices =
+            Kokkos::subview(localColumnIndices, Kokkos::make_pair(colStart, colStart + numCols));
+        auto rowValues = Kokkos::subview(values, Kokkos::make_pair(colStart, colStart + numCols));
+        Kokkos::Experimental::sort_by_key_team(t, rowColumnIndices, rowValues);
+      });
+
+  auto rowPtr = Kokkos::create_mirror_view(prefixSumEntriesPerRow);
+  Kokkos::deep_copy(rowPtr, prefixSumEntriesPerRow);
+  auto valuesHost = Kokkos::create_mirror_view(values);
+  Kokkos::deep_copy(valuesHost, values);
```

Comment:
Why do the values need to be copied to host?

Reply:
Whoops -- I thought this data had to be host accessible with the particular `KokkosSparse::CrsMatrix` ctor I was using, but I think you're right that copying to the host shouldn't be necessary.